flit/validate.py

```diff
@@ -1,38 +1,49 @@
 """Validate various pieces of packaging data"""
 
+import functools
 import io
 import logging
 import os
 from pathlib import Path
 import re
 import requests
 import sys
+import tempfile
 
 from .vendorized.readme.rst import render
 
 log = logging.getLogger(__name__)
 
+@functools.lru_cache(maxsize=1)
 def get_cache_dir() -> Path:
     """Locate a platform-appropriate cache directory for flit to use
 
-    Does not ensure that the cache directory exists.
+    Ensures that the cache directory exists.
     """
     # Linux, Unix, AIX, etc.
     if os.name == 'posix' and sys.platform != 'darwin':
         # use ~/.cache if empty OR not set
-        xdg = os.environ.get("XDG_CACHE_HOME", None) \
+        local = os.environ.get("XDG_CACHE_HOME", None) \
               or os.path.expanduser('~/.cache')
-        return Path(xdg, 'flit')
 
     # Mac OS
     elif sys.platform == 'darwin':
-        return Path(os.path.expanduser('~'), 'Library/Caches/flit')
+        local = os.path.expanduser('~/Library/Caches')
 
     # Windows (hopefully)
     else:
         local = os.environ.get('LOCALAPPDATA', None) \
                 or os.path.expanduser('~\\AppData\\Local')
-        return Path(local, 'flit')
+
+    cache_dir = Path(local, "flit")
+    try:
+        cache_dir.mkdir(parents=True)
+    except FileExistsError:
+        pass
+    except (OSError, PermissionError):
+        cache_dir = Path(tempfile.TemporaryDirectory().name)
+        cache_dir.mkdir(parents=True)
```

Comment:
AFAIK, `TemporaryDirectory()` always creates the directory, so we don't need to do so here.

Reply:
I thought so too, but in one of my test setups this was not working reliably. Since I'm not at home and can't remember by heart which one it was,  I'd need to check on this later.

Reply:
Aha, I think I've got it. `TemporaryDirectory()` cleans itself up when the object is destroyed - which is immediately, because the code doesn't keep a reference to it. Use [mkdtemp](https://docs.python.org/3/library/tempfile.html#tempfile.mkdtemp) instead to get a directory which won't clear itself up.

Of course, this does mean each run of flit will create a new temp directory and not clear it up, which isn't ideal.

Reply:
m(
Thanks for pointing this out. I ran head over heels into this kind of race condition in the first place and made things worse afterwards. Feeling kind of stupid now ;-).
Do you have any idea how to solve this without falling back on a module scope variable?

Reply:
Unfortunately adding a new attribute to a pathlib.Path instance doesn't work.

Reply:
To do what you're currently doing, with a temporary directory that never gets cleaned up, the easiest thing is to use the `mkdtemp()` function instead of `TemporaryDirectory()`.

Reply:
The other possibility would be - as you described - to use mkdtemp directly, but the cleanup afterwards must also be managed somehow.

Reply:
Sorry for spamming you with comments. 
I think I have a fix in commit bef5f2b using a context manager. I had to slightly rewrite validate_classifiers also. #

Reply:
@takluyver Sorry, I didn't see your reply – my connection is a little bit flaky, sitting on a train today. I actually want an automatic cleanup. Circumventing it was just a stupid mistake on my side.